File: app/src/main/java/com/example/kotlin/dualsim/TelephonyManagement.java
package com.example.kotlin.dualsim;

import android.content.Context;
import android.util.Log;


public class TelephonyManagement {

    private static final String TAG = "chip";

    public static class TelephonyInfo {

        String imeiSIM1 = "";

        String imeiSIM2 = "";


        int stateSIM1 = 1;

        int stateSIM2 = 1;





        public String getImeiSIM1() {
            return imeiSIM1;
        }

        void setImeiSIM1(String imeiSIM1) {
            if (imeiSIM1 != null) {
                this.imeiSIM1 = imeiSIM1;
            }
        }

        public String getImeiSIM2() {
            return imeiSIM2;
        }

        void setImeiSIM2(String imeiSIM2) {
            if (imeiSIM2 != null) {
                this.imeiSIM2 = imeiSIM2;
            }
        }

        public int getStateSIM1() {
            return stateSIM1;
        }

        void setStateSIM1(int state) {
            stateSIM1 = state;
        }

        public int getStateSIM2() {
            return stateSIM2;
        }

        void setStateSIM2(int state) {
            stateSIM2 = state;
        }


    }

    private TelephonyManagement() {
    }

    public static TelephonyManagement getInstance() {
        if (mInstance == null) {
            mInstance = new TelephonyManagement();
        }
        return mInstance;
    }

    private static TelephonyManagement mInstance;

    private static DualsimBase mDualsimChip;

    /**
     * Get telephony SIM information
     * request permission {@link android.Manifest.permission#READ_PHONE_STATE}
     *
     * @param context 上下文
     * @return TelephonyInfo object
     */
    public TelephonyInfo getTelephonyInfo(Context context) {
        DualsimBase dualsim = getDualSimChip(context);
        if (dualsim.getTelephonyInfo() == null) {
            dualsim.update(context);
        }
        return dualsim.getTelephonyInfo();
    }

    /**
     * Reload telephonyinfo
     * * request permission {@link android.Manifest.permission#READ_PHONE_STATE}
     *
     * @param context 上下文
     * @return this
     */
    public TelephonyManagement updateTelephonyInfo(Context context) {

        getDualSimChip(context).update(context);

        return this;
    }

    /**
     * * request permission {@link android.Manifest.permission#READ_PHONE_STATE}
     *
     * @param context 上下文
     * @return DualsimBase object
     */
    public DualsimBase getDualSimChip(Context context) {

        if (mDualsimChip != null) {
            return mDualsimChip;
        }


        //samsung chip
        if (SamsungDualSim.getInstance(context).isSamsungDualSystem()) {
            if (SamsungDualSim.getInstance(context).getSimState(DualsimBase.TYPE_SIM_MAIN) == 0
                    && SamsungDualSim.getInstance(context).getSimState(DualsimBase.TYPE_SIM_ASSISTANT) == 0) {
                Log.w(TAG, ">>>>>>>>>use normal chip<<<<<<<<<<<");
                return mDualsimChip = NormalDualSim.getInstance(context);
            }
            Log.w(TAG, ">>>>>>>>>use samsung chip<<<<<<<<<<<");
            return mDualsimChip = SamsungDualSim.getInstance(context);
        }

        //MTK chip
        if (MTKDualSim.getInstance(context).isMTKSystem()) {
            Log.w(TAG, ">>>>>>>>>use MTK chip<<<<<<<<<<<");
            return mDualsimChip = MTKDualSim.getInstance(context);
        }

        //qualcomm chip
        if (QualcommDualSim.getInstance(context).isQualcommSystem(context)) {
            Log.w(TAG, ">>>>>>>>>use qualcomm chip<<<<<<<<<<<");
            return mDualsimChip = QualcommDualSim.getInstance(context);
        }

        //normal chip
        Log.w(TAG, ">>>>>>>>>use normal chip<<<<<<<<<<<");
        return mDualsimChip = NormalDualSim.getInstance(context);
    }

}

File: app/src/main/java/com/example/kotlin/dualsim/TelephonyManagement.kt
package com.example.kotlin.dualsim

import android.content.Context


class TelephonyManagement private constructor() {
    
    /**
     * Reload telephonyinfo
     * * request permission [android.Manifest.permission.READ_PHONE_STATE]
     *
     * @param context 上下文
     * @return this
     */
    fun updateTelephonyInfo(context: Context?): TelephonyManagement {
        getDualSimChip(context)!!.update(context)
        return this
    }

    /**
     * * request permission [android.Manifest.permission.READ_PHONE_STATE]
     *
     * @param context 上下文
     * @return DualsimBase object
     */
    private fun getDualSimChip(context: Context?): DualSimBase? {
        if (mDualsimChip != null) {
            return mDualsimChip
        }
        //samsung chip
        if (SamsungDualSim.getInstance(context!!)!!.isSamsungDualSystem) {
            if (SamsungDualSim.getInstance(context)!!.getSimState(DualSimBase.TYPE_SIM_MAIN) == 0
                && SamsungDualSim.getInstance(context)!!.getSimState(DualSimBase.TYPE_SIM_ASSISTANT) == 0
            ) {
                return NormalDualSim.getInstance(context).also { mDualsimChip = it }
            }
            return SamsungDualSim.getInstance(context).also { mDualsimChip = it }
        }

        //MTK chip
        if (MTKDualSim.getInstance(context)!!.isMTKSystem) {
            return MTKDualSim.getInstance(context).also { mDualsimChip = it }
        }

        //qualcomm chip
        if (QualcommDualSim.getInstance(context)!!.isQualcommSystem()) {
            return QualcommDualSim.getInstance(context).also { mDualsimChip = it }
        }

        //normal chip
        return NormalDualSim.getInstance(context).also { mDualsimChip = it }
    }

    companion object {
        val instance: TelephonyManagement?
            get() {
                if (mInstance == null) {
                    mInstance = TelephonyManagement()
                }
                return mInstance
            }
        private var mInstance: TelephonyManagement? = null
        private var mDualsimChip: DualSimBase? = null
    }

    class TelephonyInfo{

        var imeiSIM1 = ""

        var imeiSIM2 = ""


        private var stateSIM1 = 1

        private var stateSIM2 = 1


        @JvmName("setImeiSIM11")
        fun setImeiSIM1(imeiSIM1: String?) {
            if (imeiSIM1 != null) {
                this.imeiSIM1 = imeiSIM1
            }
        }

        @JvmName("setImeiSIM21")
        fun setImeiSIM2(imeiSIM2: String?) {
            if (imeiSIM2 != null) {
                this.imeiSIM2 = imeiSIM2
            }
        }

        @JvmName("getStateSIM11")
        fun getStateSIM1(): Int {
            return stateSIM1
        }

        @JvmName("setStateSIM11")
        fun setStateSIM1(state: Int) {
            stateSIM1 = state
        }

        @JvmName("getStateSIM21")
        fun getStateSIM2(): Int {
            return stateSIM2
        }

        @JvmName("setStateSIM21")
        fun setStateSIM2(state: Int) {
            stateSIM2 = state
        }


    }

    /**
     * Get telephony SIM information
     * request permission [android.Manifest.permission.READ_PHONE_STATE]
     *
     * @param context 上下文
     * @return TelephonyInfo object
     */
    fun getTelephonyInfo(context: Context?): TelephonyInfo? {
        val dualsim = getDualSimChip(context)
        if (dualsim!!.telephonyInfo == null) {
            dualsim.update(context)
        }
        return dualsim.telephonyInfo
    }
}